### ISSUE TITLE
Insufficient control over RUNNING_IN_VALGRIND, should only be active with --tool=memcheck

### ISSUE THREAD

**Issue by stefvanvlierberghe** [gperftools/gperftools#1167]
gperftools-2.7 

When running in valgrind with --tool=callgrind or --tool=none the tcmalloc library is still disabling the function calls to the malloc extensions although with these tools the malloc calls are not intercepted by valgrind.

Current control is this :

static int GetRunningOnValgrind(void) {
#ifdef RUNNING_ON_VALGRIND
  if (RUNNING_ON_VALGRIND) return 1;
#endif
  const char *running_on_valgrind_str = TCMallocGetenvSafe("RUNNING_ON_VALGRIND");
  if (running_on_valgrind_str) {
    return strcmp(running_on_valgrind_str, "0") != 0;
  }
  return 0;
}

This means that if the library was built with -DRUNNING_ON_VALGRIND then the dynamic check for running on valgrind makes a (final) decision, and those who set an env variable RUNNING_ON_VALGRIND can only influence this decision when built without -DRUNNING_ON_VALGRIND.

The build with -DRUNNING_ON_VALGRIND is definitely better to avoid that people that are unaware of the config get puzzled by the valgrind intercepts breaking the malloc extensions, but it is worse in the sense that is takes away all control over the behaviour for those who are aware of the situation.

As explained above, this implies that in the former case one can never measure performance of the tcmalloc library and still output heap usage statistics in this mode.

There is a trivial fix : revert the two conditions such that those who set the environment variable will always have full control.

If you want to be fully compatible with the old behaviour (in case some people would count on the environment variable having no effect) a slightly larger fix is to introduce another variable, like this:

static int GetRunningOnValgrind(void) {
  const char *force_running_on_valgrind_str = TCMallocGetenvSafe("FORCE_RUNNING_ON_VALGRIND");
  if (force_running_on_valgrind_str) {
    return strcmp(force_running_on_valgrind_str, "0") != 0;
  }
#ifdef RUNNING_ON_VALGRIND
  if (RUNNING_ON_VALGRIND) return 1;
#endif
  const char *running_on_valgrind_str = TCMallocGetenvSafe("RUNNING_ON_VALGRIND");
  if (running_on_valgrind_str) {
    return strcmp(running_on_valgrind_str, "0") != 0;
  }
  return 0;
}

But in my opinion it would be unlikely that people would set RUNNING_ON_VALGRIND=0 with a different intent than preserving the malloc extensions (i.e. unlikely this would be used in combination with --tool=memcheck or --tool=massif).

These are just the two most trivial ways to solve the problem, possibly there are better ways, but current behavior is not ideal and there is currently no way to dynamically check if valgrind is replacing the heap interface, so I would recommend to simply swap the two conditions.

All the best,
Stef

**Comment by alk** [comment]
I am not aware of people building tcmalloc with -DRUNNING_ON_VALGRIND. Maybe couple decades ago Google builds could define it. We could perhaps simply drop this define.

Otherwise, your proposal looks sensible. So feel free to send patch.

**Comment by stefvanvlierberghe** [comment]
Thanks, the patch below works ok for me, RUNNING_ON_VALGRIND=0 can now override the auto-detect when using --tool=callgrind :

--- src/base/dynamic_annotations.c.orig Thu Aug 25 05:12:24 2016
*** src/base/dynamic_annotations.c      Mon Feb 10 20:13:32 2020
***************
*** 138,150 ****
  #if DYNAMIC_ANNOTATIONS_EXTERNAL_IMPL == 0
  
  static int GetRunningOnValgrind(void) {
- #ifdef RUNNING_ON_VALGRIND
-   if (RUNNING_ON_VALGRIND) return 1;
- #endif
    const char *running_on_valgrind_str = TCMallocGetenvSafe("RUNNING_ON_VALGRIND");
    if (running_on_valgrind_str) {
      return strcmp(running_on_valgrind_str, "0") != 0;
    }
    return 0;
  }
  
--- 138,150 ----
  #if DYNAMIC_ANNOTATIONS_EXTERNAL_IMPL == 0
  
  static int GetRunningOnValgrind(void) {
    const char *running_on_valgrind_str = TCMallocGetenvSafe("RUNNING_ON_VALGRIND");
    if (running_on_valgrind_str) {
      return strcmp(running_on_valgrind_str, "0") != 0;
    }
+ #ifdef RUNNING_ON_VALGRIND
+   if (RUNNING_ON_VALGRIND) return 1;
+ #endif
    return 0;
  }

**Comment by paulfloyd** [comment]
OK. DHAT, DRD, Helgrind, Massif and Memcheck all replace malloc. The other tools don't - Callgrind, Cachegrind, Lackey, Exp-bbv and of course None.

Maybe I could add a VALGRIND_REPLACES_MALLOC macro that you could use?

**Comment by paulfloyd** [comment]
There is no easy answer to this.

The problem with using an environment variable is that most users are not aware of its existence. 

That leaves some sort of automatic detection. The problem with that is package maintainers. In my experience package maintainers will avoid adding optional dependencies at all costs. I doubt that any gperftools binary package gets built with a dependency on the Valgrind headers. That means that someone that uses apt/dnf/pkg or whatever won't get gperftools with automatic detection. I imagine that it is because of this that gperftools added the environment variable as a backup mechanism.

From my perspective the best answer would be for gperftools to copy valgrind.h (and it is specifically has a permissive license to allow that). The file doesn't change that often, roughly once a year (the next change that I anticipate is for macOS arm64 support some time this year, with a bit of luck).

**Comment by alk** [comment]
Yes, definitely okay. Both are acceptable, though. Thanks for taking a look at this.
